File: week_01/p_set_01/credit.c
#include <cs50.h>
#include <math.h>
#include <stdio.h>

void validate_card(long card_number);

int main(void)
{
    long card_number;
    card_number = get_long("Number: ");
    validate_card(card_number);
}

void validate_card(long card_number)
{
    long og_card_number = card_number;
    int digit;
    int digit_sum = 0;

    int i = 0;
    while (card_number)
    {
        // Get single digit counting backwards
        digit = card_number % 10;

        // Using a counter, get every other one, from second to last and double
        // it
        if (i % 2 != 0)
        {
            digit *= 2;
        }

        // Sum the digits
        digit_sum += digit % 10;
        digit_sum += digit / 10;

        // Remove that last digit and count up
        card_number /= 10;
        i++;
    }
    // Check its validity first parameter
    if (digit_sum % 10 != 0)
    {
        printf("INVALID\n");
    }
    else
    {
        // Store initial digits to test the card company
        long two_initial_digits = og_card_number / (pow(10, i - 2));
        long first_digit = og_card_number / (pow(10, i - 1));

        // Test against companies indicators
        if (i == 13 && first_digit == 4)
        {
            printf("VISA\n");
        }
        else if (i == 15)
        {
            if (two_initial_digits == 34 || two_initial_digits == 37)
            {
                printf("AMEX\n");
            }
            else
            {
                printf("INVALID\n");
            }
        }
        else if (i == 16)
        {
            if (first_digit == 4)
            {
                printf("VISA\n");
            }
            else if (two_initial_digits == 51 || two_initial_digits == 52 ||
                     two_initial_digits == 53 || two_initial_digits == 54 ||
                     two_initial_digits == 55)
            {
                printf("MASTERCARD\n");
            }
            else
            {
                printf("INVALID\n");
            }
        }
        else
        {
            printf("INVALID\n");
        }
    }
}

File: weeks/week_06/p_set_06/credit.py
from cs50 import get_string
import re
import sys


def main():
    card = get_string("Number: ")

    card_company = match_digits_length(card)
    if card_company:
        if validate_card_digits(card):
            print(card_company)
            sys.exit(0)

    print("INVALID")
    sys.exit(1)


card_companies = [
    {
        "name": "VISA",
        "regex": re.compile("^4\\d{12}$")
    },
    {
        "name": "VISA",
        "regex": re.compile("^4\\d{15}$")
    },
    {
        "name": "AMEX",
        "regex": re.compile("^34|37\\d{13}$")
    },
    {
        "name": "MASTERCARD",
        "regex": re.compile("^5[1-5]\\d{14}$")
    },


]


def match_digits_length(card):
    for company in card_companies:
        card_match = re.search(company['regex'], card)
        if card_match:
            return company['name']

    return False


def validate_card_digits(card_number):
    card = int(card_number)
    digit, counter, digit_sum = 0, 0, 0

    while card:
        # Get single digit counting backwards
        digit = card % 10

        # Using a counter, get every other one, from the second and double it
        if counter % 2 != 0:
            digit *= 2

        # Sum the digits
        digit_sum += digit % 10
        digit_sum += digit // 10

        # Remove that last digit and count up
        card //= 10
        counter += 1

    if digit_sum % 10 == 0:
        return True

    return False


if __name__ == "__main__":
    main()
